src/Catalyst.Core.Modules.Kvm/DeltaExecutor.cs

```diff
@@ -0,0 +1,604 @@
+                Execute(contractEntry, stateUpdate, txTracer, readOnly);
+            }
+
+            // revert state if any fails (take snapshot)
+            foreach (ContractEntry contractEntry in delta.ContractEntries)
+            {
+                Execute(contractEntry, stateUpdate, txTracer, readOnly);
+            }
+
+            if (!readOnly)
+            {
+                // we should assign block rewards here (or in Ledger)
+                _stateProvider.CommitTree();
+                _storageProvider.CommitTrees();
+            }
+            else
+            {
+                _storageProvider.Reset();
+                _stateProvider.Reset();
+            }
+        }
+
+        private void TraceLogInvalidTx(ContractEntry entry, string reason)
+        {
+            if (_logger.IsEnabled(LogEventLevel.Verbose))
+            {
+                _logger.Verbose("Invalid enrty {entry} ({reason})", entry, reason);
+            }
+        }
+
+        /// <summary>
+        /// Executes the <paramref name="entry"/>.
+        /// </summary>
+        /// <param name="entry">Transaction entry to be executed inside the <see cref="VirtualMachine"/>.</param>
+        /// <param name="stateUpdate"><see cref="Delta"/> to be used for execution environment construction</param>
+        /// <param name="txTracer">Tracer to extract the execution steps for debugging or analytics.</param>
+        /// <param name="readOnly">Defines whether the state should be reverted after the execution.</param>
+        /// <exception cref="TransactionCollisionException">Thrown when deployment address already has some code.</exception>
+        /// <exception cref="OutOfGasException">Thrown when not enough gas is available for deposit.</exception>
+        private void Execute(ContractEntry entry, StateUpdate stateUpdate, ITxTracer txTracer, bool readOnly)
+        {
+            IReleaseSpec spec = _specProvider.GetSpec(stateUpdate.Number);
+
+            (Address sender, Address recipient) = ExtractSenderAndRecipient(entry);
+            bool isPrecompile = recipient.IsPrecompiled(spec);
+            ExecutionEnvironment env = PrepareEnv(entry, sender, recipient, stateUpdate, isPrecompile);
+
+            ulong gasLimit = entry.GasLimit;
+            ulong intrinsicGas = CalculateIntrinsicGas(entry, spec);
+
+            if (_logger.IsEnabled(LogEventLevel.Verbose))
+            {
+                _logger.Verbose("Executing entry {entry}", entry);
+            }
+
+            if (!ValidateSender(entry, env, txTracer))
+            {
+                return;
+            }
+
+            if (!ValidateIntrinsicGas(entry, env, intrinsicGas, txTracer))
+            {
+                return;
+            }
+
+            if (!ValidateDeltaGasLimit(entry, env, txTracer))
+            {
+                return;
+            }
+
+            if (!_stateProvider.AccountExists(env.Sender))
+            {
+                if (env.GasPrice == UInt256.Zero)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Merge this if statement with the enclosing one.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4350015)